win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+                    /*Codes_SRS_FILE_WIN32_43_024: [ If WriteFile succeeds synchronously then file_write_async shall succeed, call CancelThreadpoolIo, call user_callback and return FILE_WRITE_ASYNC_OK.]*/
+                    CancelThreadpoolIo(handle->ptp_io);
+                    user_callback(user_context, true);
+                    result = FILE_WRITE_ASYNC_OK;
+                }
+                if (!callback_will_be_called)
+                {
+                    if (!CloseHandle(io->ov.hEvent))
+                    {
+                        LogError("Failure in CloseHandle");
+                    }
+                }
+            }
+            if (!callback_will_be_called)
+            {
+                free(io);
+            }
+        }
+    }
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, unsigned char*, destination, uint32_t, size, uint64_t, position, FILE_CB, user_callback, void*, user_context)
+{
+    FILE_READ_ASYNC_RESULT result;
+    if
+    (
+        /*Codes_SRS_FILE_43_017: [ If handle is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_046: [ If handle is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        (handle == NULL) ||
+        /*Codes_SRS_FILE_43_032: [ If destination is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_051: [ If destination is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        (destination == NULL) ||
+        /*Codes_SRS_FILE_43_020: [ If user_callback is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_047: [ If user_callback is NULL then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        (user_callback == NULL) ||
+        /*Codes_SRS_FILE_43_043: [ If size is 0 then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        /*Codes_SRS_FILE_WIN32_43_062: [ If size is 0 then file_read_async shall fail and return FILE_READ_ASYNC_INVALID_ARGS. ]*/
+        (size == 0)
+    )
+    {
+        LogError("Invalid arguments to file_read_async: FILE_HANDLE file_handle=%p, unsigned char* destination=%p, uin32_t size=%" PRIu32 ", uint64_t position=%" PRIu64 ", FILE_CB user_callback=%p, void* user_context=%p",
+            handle, destination, size, position, user_callback, user_context);
+        result = FILE_READ_ASYNC_INVALID_ARGS;
+    }
+    else
+    {
+        bool callback_will_be_called = false;
+        /*Codes_SRS_FILE_WIN32_43_026: [ file_read_async shall allocate a context to store the allocated OVERLAPPED struct, destination, handle, size, user_callback and user_context ]*/
+        FILE_WIN32_IO* io = malloc(sizeof(FILE_WIN32_IO));

```

Comment:
>io [](start = 23, length = 2)

nit: I'd consider naming this `io_context` or something just because later in the callback you have a parameter named `io` that is different. Same for the write_file of course

Reply:
done

---
In reply to: [456709491](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456709491) [](ancestors = 456709491)